dart-impl/tasking/src/internal/dart_tasking_instrumentation.c

```diff
@@ -1,47 +1,317 @@
+{
+    
+    if (dart_tool_task_finalize_cb_data.cb) {
+        dart_tool_task_finalize_cb_data.cb(dart_tool_task_finalize_cb_data.userdata);
+    }
 }
+
+void dart__tasking__instrument_local_dep(
+    dart_task_t *task1,
+    dart_task_t *task2,
+    uint64_t memaddr_raw,
+    int32_t unitid,
+    int edge_type)
+{
+    if (dart_tool_local_dep_cb_data.cb) {
+        dart_tool_local_dep_cb_data.cb((uint64_t) task1,
+                                       (uint64_t) task2,
+                                       memaddr_raw,
+                                       unitid,
+                                       edge_type,
+                                       dart_tool_local_dep_cb_data.userdata);
+    }
+}
+
+void dart__tasking__instrument_task_add_to_queue(
+    dart_task_t *task,
+    dart_thread_t *thread)
+{
+    if (dart_tool_task_add_to_queue_cb_data.cb) {
+        dart_tool_task_add_to_queue_cb_data.cb((uint64_t) task,
+                                               (uint64_t) thread,
+                                               dart_tool_task_add_to_queue_cb_data.userdata);
+    }
+}
+
+void dart__tasking__instrument_remote_dep(
+    uint64_t local_task,
+    uint64_t remote_task,
+    int local_dep_type,
+    int remote_dep_type,
+    uint64_t memaddr,
+    int32_t local_unitid,
+    int32_t remote_unitid,
+    int edge_type)
+{
+    if (dart_tool_remote_dep_cb_data.cb) {
+        dart_tool_remote_dep_cb_data.cb(local_task,
+                                        remote_task,
+                                        local_dep_type,
+                                        remote_dep_type,
+                                        memaddr,
+                                        local_unitid,
+                                        remote_unitid,
+                                        edge_type,
+                                        dart_tool_remote_dep_cb_data.userdata);
+    }
+}
+void dart__tasking__init_tools_interface(){
+  void *handle;
+  int (*toolinit)(int, int, int);
+  int toolhandle;
+  /**
+   * The name of the environment variable containing the path to the tool is stored in 
+   * DART__TOOLS_TOOL_ENV_VAR_PATH
+  */
+  const char* var = dart__base__env__string(DART__TOOLS_TOOL_ENV_VAR_PATH);
+  if (!var) {
+    //do nothing
+    DART_LOG_WARN("Tool interface disabled on unit %d.", myguid.id);
+  } else if (*var == '\0') {
+      DART_LOG_ERROR("Environment variable is an empty string!");
+  } else {
+      DART_LOG_TRACE("DART_TOOL_PATH=%s", var);
+      handle = dlopen(var, RTLD_LAZY);
+      if (!handle) {
+        /* failed to load the tool */
+        DART_LOG_TRACE("Failed to load the tool");
+        fprintf(stderr, "Error: %s\n", dlerror());
+      }
+      /**
+       * The init function name has to be stored in DART__TOOLS_TOOL_INIT_FUNCTION_NAME
+       * in dart_tools.h 
+      */
+      *(int **)(&toolinit) = dlsym(handle, DART__TOOLS_TOOL_INIT_FUNCTION_NAME);
```

Comment:
Needed?

Reply:
Yes (lvalue required as left operand of assignment).